hack/generator/spec/case-study-rolling-storage-versions.md

```diff
@@ -0,0 +1,934 @@
+    MailingAddress *Address // *** Was mandatory, now optional ***
+}
+```
+
+## Storage Conversion
+
+The storage versions are identical to those used previously and are not shown here.
+
+What does change is the `ConvertToStorage()` method, which now needs to handle the case where the `MailingAddress` has not been included:
+
+``` go
+package v20170707
+
+import storage "v20170707storage"
+
+// ConvertTo converts this Person to the Hub version.
+func (person *Person) ConvertToStorage(dest storage.Person) error {
+    dest.SortKey = person.AlphaKey
+    dest.FamilyName = person.FamilyName
+    dest.Id = person.Id
+    dest.KnownAs = person.KnownAs
+    dest.LegalName = person.LegalName
+
+    // *** Need to check whether we have a mailing address to copy ***
+    if person.MailingAddress != nil {
+        address := &storage.Address{}
+        err := person.MailingAddress.ConvertToStorage(address)
+        if err != nil {
+            return err
+        }
+
+        dest.MailingAddress = address
+    }
+
+    if assignable, ok := person.(AssignableToPerson); ok {
+        err := assignable.AssignTo(dest)
+        if err != nill {
+            return err
+        }
+    }
+
+    return nil
+}
+```
+
+If we instead had an _optional_ field that became _required_ in a later version of the API, the generated code for `ConvertToStorage()` would become simpler as the check for **nil** would not be needed.
+
+## Version Map
+
+![](images/case-study-rolling-storage-2017-07-07.png)
+
+## How often does optionality change?
+
+At the time of writing, there are **100** version-to-version changes where fields became **optional** in the later version of the API, and **99** version-to-version changes where fields became **required**.
+
+# Version 2018-08-08 - Extending nested properties
+
+Defining an address simply as `Street` and `City` has been found to be overly simplistic, so this release makes changes to allow a more flexible approach.
+
+``` go
+package v20180808
+
+type Address struct {
+    // FullAddress shows the entire address as should be used on postage
+    FullAddress  string
+    City         string
+    Country      string
+    PostCode     string
+}
+```
+
+As before, the storage version changes to match, with prior conversions using the property bag to store additional properties:
+
+``` go
+package v20180808storage 
+
+type Address struct {
+    PropertyBag
+    City         *string
+    Country      *string
+    FullAddress  *string
+    PostCode     *string
+}
+```
+
+These changes are entirely similar to those previously covered in version 2014-04-04, above.
+
+## Version Map
+
+In this release, we see that support for both `2014-04-04` and the preview version `2014-04-04preview` has been dropped:
+
+![](images/case-study-rolling-storage-2018-08-08.png)
+
+Users still running earlier releases of the service operator that are using `2014-04-04` or earlier will need to install an intermediate release in order to upgrade to this one.
```

Comment:
This could be a poor experience for people who find themselves in this situation. We should think about ways we can mitigate or simplify this (or retain all conversions).

Reply:
Agreed - it comes down to answering the question "How long do we retain older storage versions?". We would want at least a year, as a lower bound. We might want to consider keeping them indefinitely - though there's not a zero cost to doing so.

Reply:
Capture as a separate issue for later discussion; outside of the scope of this spec.